keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,167 @@
+We must provide guarantee in both of the following aspects:
+- Retain request memory to reduce allocation latency
+- Protect request memory from being reclaimed
+
+#### Node Availability
+The stability of kubelet node is very important to users. As the key resource of the node, the availability of memory is the key factor for the stability of the node. We should do something to protect node reserved memory.
+
+### Comparison with Memory Manager
+The Memory Manager is a new component of kubelet ecosystem proposed to enable single-NUMA and multi-NUMA guaranteed memory allocation at topology level. And this Memory QoS proposal mainly uses the cgrups v2 mechanism to improve the quality of memory allocation, thereby improving the memory qos of guanranteed and burstable pods and even entire node. 
+See also https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager
+
+## Design Details
+### Feature Gate
+To set `--feature-gates=MemoryQoS=true` to enable the feature.
+
+### Rules
+#### memory.low vs memory.min
+In cgroups v2, `memory.low` is designed for best-effort memory protection which is more like "soft guarantee" and won't be reclaimed unless memory can't be reclaimed from any unprotected cgroups. `memory.min` is a bit aggressive. It will always retain specified amount of memory and it can be never reclaimed. When requirement is not satisfied, system OOM killer will be invoked. 
+
+#### Mapping Details
+1. If container sets `requests.memory`, we set `memory.min=requests.memory` at container level cgroup
+2. If any containers in pod sets `requests.memory`, we set `memory.min=sum(pod.spec.containers.resources.requests[memory])` at pod level cgroup
+3. If kubelet sets `kube-reserved` and `kube-reserved-cgroup`, we set `memory.min=kube-reserved` at `kube-reserved-cgroup` node level cgroup
+4. If kubelet sets `system-reserved` and `system-reserved-cgroup`, we set `memory.min=system-reserved` at `system-reserved-cgroup` node level cgroup
+5. There are no changes regarding memory limit, that is `memory.max=memory limits` (same as existing cgroup v2 implementation)
+
+### Interactive
+New `Unified` field will be added in both CRI and QoS Manager for cgroups v2 extra parameters. It is recommended to has same semantics with opencontainers/runtime-spec#1040
+
+- container level: `Unified` added in `LinuxContainerResources`
+- pod level: `Unified` added in `cm.ResourceConfig` 
+- node level: `Unified` added in `cm.ResourceConfig` 
+
+### Workflow
+#### Container Level Cgroup
+![](./container-level.png)
+
+#### Pod Level Cgroup
+![](./pod-level.png)
+
+#### Node Level Cgroup
+![](./node-level.png)
+
+### Cgroup Hierarchy 
+Pod cgroup:
+```
+/cgroup2/kubepods/pod<UID>/memory.min=sum(pod.spec.containers.resources.requests[memory])
+```
+
+Container cgroup:
+```
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.min=pod.spec.containers[i].resources.requests[memory]
```

Comment:
As I understand it, `/cgroup2/kubepods` also have to get a `memory.min` value, the sum of all pods, not only the pod/container cgroups. This will therefore apply to `/cgroup2/kubepods/kubepods-<qos>` in case that is enabled as well. Or am I wrong?

Linux docs says:
> Effective min boundary is limited by memory.min values of all ancestor cgroups.


Also note the cgroup2 fs mount option:


*memory_recursiveprot*: Recursively apply memory.min and memory.low protection to entire subtrees, without requiring explicit downward propagation into leaf cgroups. This allows protecting entire subtrees from one another, while retaining free competition within those subtrees. This should have been the default behavior but is a mount-option to avoid regressing setups relying on the original semantics (e.g. specifying bogusly high ‘bypass’ protection values at higher tree levels).

Reply:
Yes, you are right. I have added qos level design for that. Thanks.